NPE when retrying RPC requests with armeria-client (suspect race condition around `NonWrappingRequestContext#rpcReq`)

Hi,

We're seeing exceptions like:

````
io.grpc.StatusRuntimeException: UNKNOWN
        at io.grpc.Status.asRuntimeException(Status.java:533)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
        at com.linecorp.armeria.internal.client.grpc.ArmeriaClientCall.close(ArmeriaClientCall.java:464)
        at com.linecorp.armeria.internal.client.grpc.ArmeriaClientCall.transportReportStatus(ArmeriaClientCall.java:416)
        at com.linecorp.armeria.internal.common.grpc.TransportStatusListener.transportReportStatus(TransportStatusListener.java:28)
        at com.linecorp.armeria.internal.common.grpc.HttpStreamDeframerHandler.processOnError(HttpStreamDeframerHandler.java:124)
        at com.linecorp.armeria.common.stream.DefaultHttpDeframer.onError0(DefaultHttpDeframer.java:281)
        at com.linecorp.armeria.common.stream.DefaultHttpDeframer.onError(DefaultHttpDeframer.java:272)
        at com.linecorp.armeria.common.stream.DeferredStreamMessage$ForwardingSubscriber.onError(DeferredStreamMessage.java:346)
        at com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:281)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent0(DefaultStreamMessage.java:264)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:256)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:436)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:368)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.doRequest(DefaultStreamMessage.java:239)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.request(DefaultStreamMessage.java:224)
        at com.linecorp.armeria.common.stream.AbstractStreamMessage$SubscriptionImpl.request(AbstractStreamMessage.java:224)
        at com.linecorp.armeria.common.stream.DeferredStreamMessage$ForwardingSubscriber.onSubscribe(DeferredStreamMessage.java:334)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.subscribe(DefaultStreamMessage.java:139)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$subscribe$0(DefaultStreamMessage.java:129)
        at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:502)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException: rpcReq
        at java.base/java.util.Objects.requireNonNull(Objects.java:246)
        at com.linecorp.armeria.client.DefaultClientRequestContext.<init>(DefaultClientRequestContext.java:361)
        at com.linecorp.armeria.client.DefaultClientRequestContext.newDerivedContext(DefaultClientRequestContext.java:410)
        at com.linecorp.armeria.client.retry.AbstractRetryingClient.newDerivedContext(AbstractRetryingClient.java:268)
        at com.linecorp.armeria.client.retry.RetryingClient.doExecute0(RetryingClient.java:282)
        at com.linecorp.armeria.client.retry.RetryingClient.doExecute(RetryingClient.java:241)
        at com.linecorp.armeria.client.retry.RetryingClient.doExecute(RetryingClient.java:56)
        at com.linecorp.armeria.client.retry.AbstractRetryingClient.execute(AbstractRetryingClient.java:96)
        at com.linecorp.armeria.client.limit.AbstractConcurrencyLimitingClient$PendingTask.run(AbstractConcurrencyLimitingClient.java:237)
        at com.linecorp.armeria.client.limit.AbstractConcurrencyLimitingClient.drain(AbstractConcurrencyLimitingClient.java:184)
        at com.linecorp.armeria.client.limit.AbstractConcurrencyLimitingClient$PendingTask.lambda$run$0(AbstractConcurrencyLimitingClient.java:240)
        at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930)
        at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907)
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
        ... 7 common frames omitted
````

`RetryingClient#doExecute0` passes `ctx.rpcRequest()` to `newDerivedContext`, so I think this can only happen if `ctx.rpcReq` is `null` on this line but then becomes non-null when the constructor of `DefaultClientRequestContext` is executed. However that does seem entirely possible: `rpcReq` is `volatile` and `DefaultRequestLog#requestContent` calls `ctx.updateRpcRequest` if `ctx.rpcRequest()` is `null`. `DefaultRequestLog#requestContent` is called from several places; while I don't have a specific scenario for why it would be called at the same time as `AbstractRetryingClient#execute`, there doesn't seem to be any synchronisation is place to prevent that.

Under what circumstances is it permitted for `ctx.rpcReq` to change from `null` to non-`null`, and does `newDerivedContext` handle all of those cases correctly? Or is there some other reason for this exception?

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
@@ -357,6 +357,9 @@ public final class DefaultClientRequestContext
         if (ctx.request() != null) {
             requireNonNull(req, "req");
         }
+        // The rpcReq can be null when ctx.rpcRequest() is not null because there's a chance that
+        // the rpcRequest is set between the time we call ctx.rpcRequest() to create this context and now.
+        // So we don't check the nullness of rpcRequest unlike request.
 
         eventLoop = ctx.eventLoop().withoutContext();
         options = ctx.options();

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
@@ -360,6 +360,7 @@ public final class DefaultClientRequestContext
         // The rpcReq can be null when ctx.rpcRequest() is not null because there's a chance that
         // the rpcRequest is set between the time we call ctx.rpcRequest() to create this context and now.
         // So we don't check the nullness of rpcRequest unlike request.
+        // See https://github.com/line/armeria/pull/3251 and https://github.com/line/armeria/issues/3248.
 
         eventLoop = ctx.eventLoop().withoutContext();
         options = ctx.options();

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java
@@ -357,9 +357,6 @@ public final class DefaultClientRequestContext
         if (ctx.request() != null) {
             requireNonNull(req, "req");
         }
-        if (ctx.rpcRequest() != null) {
-            requireNonNull(rpcReq, "rpcReq");
-        }
 
         eventLoop = ctx.eventLoop().withoutContext();
         options = ctx.options();

--- a/core/src/main/java/com/linecorp/armeria/client/retry/AbstractRetryingClient.java
+++ b/core/src/main/java/com/linecorp/armeria/client/retry/AbstractRetryingClient.java
@@ -276,12 +276,10 @@ public abstract class AbstractRetryingClient<I extends Request, O extends Respon
         if (parentLog.isAvailable(RequestLogProperty.NAME)) {
             final String serviceName = partial.serviceName();
             final String name = partial.name();
-            if (name != null) {
-                if (serviceName != null) {
-                    logBuilder.name(serviceName, name);
-                } else {
-                    logBuilder.name(name);
-                }
+            if (serviceName != null) {
+                logBuilder.name(serviceName, name);
+            } else {
+                logBuilder.name(name);
             }
         }
 
@@ -289,32 +287,28 @@ public abstract class AbstractRetryingClient<I extends Request, O extends Respon
         if (parentLogBuilder.isDeferred(RequestLogProperty.REQUEST_CONTENT)) {
             logBuilder.defer(RequestLogProperty.REQUEST_CONTENT);
         }
-        parentLog.whenAvailable(RequestLogProperty.REQUEST_CONTENT).thenApply(requestLog -> {
-            logBuilder.requestContent(requestLog.requestContent(), requestLog.rawRequestContent());
-            return null;
-        });
+        parentLog.whenAvailable(RequestLogProperty.REQUEST_CONTENT)
+                 .thenAccept(requestLog -> logBuilder.requestContent(
+                         requestLog.requestContent(), requestLog.rawRequestContent()));
         if (parentLogBuilder.isDeferred(RequestLogProperty.REQUEST_CONTENT_PREVIEW)) {
             logBuilder.defer(RequestLogProperty.REQUEST_CONTENT_PREVIEW);
         }
-        parentLog.whenAvailable(RequestLogProperty.REQUEST_CONTENT_PREVIEW).thenApply(requestLog -> {
-            logBuilder.requestContentPreview(requestLog.requestContentPreview());
-            return null;
-        });
+        parentLog.whenAvailable(RequestLogProperty.REQUEST_CONTENT_PREVIEW)
+                 .thenAccept(requestLog -> logBuilder.requestContentPreview(
+                         requestLog.requestContentPreview()));
 
         // Propagates the response content only when deferResponseContent is called.
         if (parentLogBuilder.isDeferred(RequestLogProperty.RESPONSE_CONTENT)) {
             logBuilder.defer(RequestLogProperty.RESPONSE_CONTENT);
-            parentLog.whenAvailable(RequestLogProperty.RESPONSE_CONTENT).thenApply(requestLog -> {
-                logBuilder.responseContent(requestLog.responseContent(), requestLog.rawResponseContent());
-                return null;
-            });
+            parentLog.whenAvailable(RequestLogProperty.RESPONSE_CONTENT)
+                     .thenAccept(requestLog -> logBuilder.responseContent(
+                             requestLog.responseContent(), requestLog.rawResponseContent()));
         }
         if (parentLogBuilder.isDeferred(RequestLogProperty.RESPONSE_CONTENT_PREVIEW)) {
             logBuilder.defer(RequestLogProperty.RESPONSE_CONTENT_PREVIEW);
-            parentLog.whenAvailable(RequestLogProperty.RESPONSE_CONTENT_PREVIEW).thenApply(requestLog -> {
-                logBuilder.responseContentPreview(requestLog.responseContentPreview());
-                return null;
-            });
+            parentLog.whenAvailable(RequestLogProperty.RESPONSE_CONTENT_PREVIEW)
+                     .thenAccept(requestLog -> logBuilder.responseContentPreview(
+                             requestLog.responseContentPreview()));
         }
         return derived;
     }

--- a/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java
+++ b/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java
@@ -768,6 +768,7 @@ final class DefaultRequestLog implements RequestLog, RequestLogBuilder {
     @Override
     public String name() {
         ensureAvailable(RequestLogProperty.NAME);
+        assert name != null;
         return name;
     }
 
@@ -908,11 +909,10 @@ final class DefaultRequestLog implements RequestLog, RequestLogBuilder {
 
         this.requestContent = requestContent;
         this.rawRequestContent = rawRequestContent;
-        updateFlags(RequestLogProperty.REQUEST_CONTENT);
-
         if (requestContent instanceof RpcRequest && ctx.rpcRequest() == null) {
             ctx.updateRpcRequest((RpcRequest) requestContent);
         }
+        updateFlags(RequestLogProperty.REQUEST_CONTENT);
 
         final int requestCompletionFlags = RequestLogProperty.FLAGS_REQUEST_COMPLETE & ~deferredFlags;
         if (isAvailable(requestCompletionFlags)) {

--- a/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientRetryTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientRetryTest.java
@@ -1,0 +1,100 @@
+package com.linecorp.armeria.client.grpc;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.ClientRequestContextCaptor;
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.client.retry.RetryRuleWithContent;
+import com.linecorp.armeria.client.retry.RetryingClient;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.grpc.GrpcSerializationFormats;
+import com.linecorp.armeria.common.grpc.protocol.GrpcHeaderNames;
+import com.linecorp.armeria.common.logging.RequestLogAccess;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleRequest;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleResponse;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc.TestServiceBlockingStub;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.grpc.GrpcService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.stub.StreamObserver;
+
+final class GrpcClientRetryTest {
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service(GrpcService.builder()
+                                  .addService(new TestServiceImpl())
+                                  .build());
+        }
+    };
+
+    @Test
+    void childrenContextsHaveSameRpcRequest() {
+        final TestServiceBlockingStub client =
+                Clients.builder(server.uri(SessionProtocol.HTTP, GrpcSerializationFormats.PROTO))
+                       .decorator(RetryingClient.newDecorator(retryRuleWithContent()))
+                       .build(TestServiceBlockingStub.class);
+
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            final SimpleResponse result = client.unaryCall(SimpleRequest.getDefaultInstance());
+            assertThat(result.getUsername()).isEqualTo("my name");
+            final ClientRequestContext context = captor.get();
+            final List<RequestLogAccess> children = context.log().children();
+            assertThat(children).hasSize(3);
+            children.forEach(child -> {
+                assertThat(context.rpcRequest()).isSameAs(child.context().rpcRequest());
+            });
+        }
+    }
+
+    private static RetryRuleWithContent<HttpResponse> retryRuleWithContent() {
+        return RetryRuleWithContent.<HttpResponse>builder()
+                .onResponseHeaders((ctx, headers) -> {
+                    // Trailers may be sent together with response headers, with no message in the body.
+                    final Integer grpcStatus = headers.getInt(GrpcHeaderNames.GRPC_STATUS);
+                    return grpcStatus != null && grpcStatus != 0;
+                })
+                .onResponse((ctx, res) -> res.aggregate().thenApply(aggregatedRes -> {
+                    final HttpHeaders trailers = aggregatedRes.trailers();
+                    return trailers.getInt(GrpcHeaderNames.GRPC_STATUS, -1) != 0;
+                }))
+                .thenBackoff();
+    }
+
+    private static class TestServiceImpl extends TestServiceGrpc.TestServiceImplBase {
+
+        private final AtomicInteger retryCounter = new AtomicInteger();
+
+        @Override
+        public void unaryCall(SimpleRequest request, StreamObserver<SimpleResponse> responseObserver) {
+            switch (retryCounter.getAndIncrement()) {
+                case 0:
+                    responseObserver.onError(new StatusException(Status.INTERNAL));
+                    break;
+                case 1:
+                    responseObserver.onNext(SimpleResponse.newBuilder().setUsername("my name").build());
+                    responseObserver.onError(new StatusException(Status.INTERNAL));
+                    break;
+                default:
+                    responseObserver.onNext(SimpleResponse.newBuilder().setUsername("my name").build());
+                    responseObserver.onCompleted();
+                    break;
+            }
+        }
+    }
+}

--- a/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientRetryTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientRetryTest.java
@@ -1,3 +1,18 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
 package com.linecorp.armeria.client.grpc;
 
 import static org.assertj.core.api.Assertions.assertThat;

